File: DynamicArray.cpp
#include "DynamicArray.h"

DynamicArray::DynamicArray(int capacity) {
    this->capacity = capacity;
    arr = new int[capacity]; // Allocate on the heap
    size = 0;
}

int DynamicArray::get(int i) {
    return arr[i];
}

void DynamicArray::set(int i, int n) {
    arr[i] = n;
}

void DynamicArray::pushback(int n) { // Add to the end of the array
    if (size == capacity) {
        resize();
    }
    arr[size++] = n;
}

int DynamicArray::popback() { // Remove from the end of the array
    return arr[--size];
}

void DynamicArray::resize() { // Double the capacity of the array
    capacity *= 2;
    int* newArr = new int[capacity];
    for (int i = 0; i < size; i++) {
        newArr[i] = arr[i];
    }
    delete[] arr;
    arr = newArr;
}

int DynamicArray::getSize() {
    return size;
}

int DynamicArray::getCapacity() {
    return capacity;
}

File: DynamicArray.cs
public class DynamicArray
{
    private int[] array;
    private int size;
    private int capacity;
    private const int DEFAULT_CAPACITY = 16;

    public DynamicArray(int capacity)
    {
        this.capacity = capacity > 0 ? capacity : DEFAULT_CAPACITY; //  Ensure positive capacity
        this.array = new int[this.capacity];
        this.size = 0;
    }

    public int Get(int i)
    {
        return this.array[i];
    }

    public void Set(int i, int n)
    {
        this.array[i] = n;
    }

    public void PushBack(int n) // Adds an element to the end
    {
        if (this.size == this.capacity)
        {
            this.Resize();
        }
        this.array[this.size++] = n;
    }

    public int PopBack() // Removes and returns the last element
    {
        return this.array[--this.size];
    }

    private void Resize() // Doubles the capacity
    {
        this.capacity *= 2;
        int[] newArray = new int[this.capacity];
        Array.Copy(this.array, newArray, this.size);
        this.array = newArray;
    }

    public int GetSize()
    {
        return this.size;
    }

    public int GetCapacity()
    {
        return this.capacity;
    }
}